### PR TITLE
20901: Add Range Based Light Attenuation

### PR DESCRIPTION
**Pull request by Geenz**
This change blends the bounds of light sources in order to have better looking lights.  Additionally, it adds a new parameter to evalLightAttenuation for the unnormalized light vector.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
An admin must verify this patch for the pull request build to fire. Type "ok to test" to begin testing this pull request or "add to whitelist" to test this and all future pull requests from this collaborator.

**Comment by cozza13** [comment]
ok to test

**Comment by zzmp** [comment]
@Geenz do you have before/after screenshots that you can add visualizing these changes?

File: libraries/model/src/model/Light.slh [hunk 1]
```diff
@@ -98,10 +98,15 @@ float getLightShowContour(Light l) {
     return l._control.w;
 }
 
-float evalLightAttenuation(Light l, float d) {
+float evalLightAttenuation(Light l, float d, vec3 ulightvec) {
```
Comment (zzmp) [inline libraries/model/src/model/Light.slh]: Can you document the parameters here? It's important that ulightvec is unnormalized, right?

Comment (Geenz) [inline libraries/model/src/model/Light.slh]: Yep.  Doing so now.

**Comment by Geenz** [comment]
These were all taken in Playa.  Test plan is as follows.

1. Log in to current version of Interface.
2. Go to Playa.
3. Find some small bright lights (such as the ones in the examples below).
4. Take screenshots.
5. Log in to the version of interface from this PR.
6. Repeat steps 2 through 4.
7. Compare screenshots.  Lights should be noticeably softer near their cut offs in this PR.


Green light with current attenuation:
screen shot 2016-05-19 at 9 00 08 pm

Green light with attenuation in this PR:
screen shot 2016-05-19 at 8 53 24 pm
Notice how where the current attenuation has a harsh edge, this now has a softer edge

Pink light with current attenuation:
screen shot 2016-05-19 at 9 02 53 pm

Pink light with attenuation in this PR:
screen shot 2016-05-19 at 9 04 38 pm
Notice that the pink spot on the ground where the light's bounds just come into contact is much less pronounced.

**Comment by samcake** [comment]
Spot on!
:)

**Comment by hifi-gustavo** [comment]
Build console output for commit: e2f5d22c9857ca4a421cf458a4f1732ff588d556 available at: https://gist.github.com/80b3c838214f61f55a4c830fbc3c981e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7945.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7945.dmg

**Comment by Nex-Pro** [comment]
I think @Geenz is going soft.

**Comment by Nex-Pro** [comment]
**PR Build: #7853**
**Testing Architecture: Desktop PC**
**RAM: 16gb**
**CPU: 4.0 ghz i7**
**GPU: Nvidia gtx770 2gb**
**Testing OS: win 10 pro** 
**Testing Locations:  playa**

#### FAIL

See screenshots:

Note: Colours look different between screenshots possibly due to changing ambient lighting or angle of view.

**PR7945 spotlight**
7945lights

**4904 spotlight**
4904lights

**Comment by Nex-Pro** [comment]
In addition the frame rates were really low and CPU usage very high.

**Comment by Geenz** [comment]
Only the radial attenuation has been modified for spot lights, so at the defined angles for them you'll still see hard edges.  However, for those lights you should still be seeing softer attenuation.  I'll investigate - performance shouldn't be measurably different either (maybe a shader isn't compiling correctly under Windows?)

**Comment by Geenz** [comment]
Can someone shed some light on why the build is failing?  I've attempted builds on both OS X and Windows without issue.

**Comment by ZappoMan** [comment]
@Geenz - the mac errors appear to be build system failures...

**Comment by ZappoMan** [comment]
test this please

**Comment by murillodigital** [comment]
This PR is failing due to warnings in OS X - these warnings were likely present in the codebase already and are now failing the build given the upgrade to Xcode in the build environment - this is not a build system failure

**Comment by ZappoMan** [comment]
@Geenz - Leo is right, you need to rebase or merge with upstream master to get the warning fixes that will make this build complete.

**Comment by Geenz** [comment]
@ZappoMan @murillodigital Understood, will do once I'm home.

**Comment by Geenz** [comment]
@ZappoMan Do you know if there's been any breaking protocol changes on master?  I can't seem to connect to Playa (or others for that matter).

**Comment by hifi-gustavo** [comment]
Build console output for commit: a80f47e5508e54eddeefe874e9eaa60bde2fea7b available at: https://gist.github.com/80b3c838214f61f55a4c830fbc3c981e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7945.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7945.dmg

**Comment by Nex-Pro** [comment]
@Geenz  Do you need me to re-test this ?

**Comment by Geenz** [comment]
@Nex-Pro I'm doing a bit of testing on this now, especially WRT the high CPU usage - that definitely should not be happening with this.  I'll let you know.

**Comment by ZappoMan** [comment]
Yes. Master will no longer connect to playa. The protocol has changed. We'll be upgrading playa when we push a new beta build. Hopefully Monday.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 180f4ba4f509e2c70d6f86905d5e4c49ecd704c9 available at: https://gist.github.com/80b3c838214f61f55a4c830fbc3c981e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7945.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7945.dmg

**Comment by Geenz** [comment]
I've tested this across WIndows, OS X, AMD, and Nvidia - all combinations seem to be giving consistent results now.  Additionally, I'm not able to consistently reproduce the high CPU usage bug - though when it does happen, I notice that a procedural shader is failing to compile (texture2d is the culprit - that's been deprecated since OpenGL 3.3 I believe?  I'll need to look up the specific shader it's complaining about - I noticed a few shaders still using texture2D).

The algorithm for this has changed slightly, you should see slightly softer looking attenuation along the edges of lights.

This is what a light colored white with a size of 5 meters will look like, with the default intensity and falloff radius:
screen shot 2016-05-22 at 9 38 37 pm

Compared to the current version:
screen shot 2016-05-22 at 9 41 56 pm

Falloff radius of 1 meter for the new one:
screen shot 2016-05-22 at 9 44 10 pm

And old:
screen shot 2016-05-22 at 9 43 13 pm

I want to revisit light attenuation later by providing an option between inverse distance squared and exponential light falloff similar to UE4, so I've created a new suggestion on WL: https://worklist.net/20923

@Nex-Pro Feel free to test this - you should see similar results now.

**Comment by hifi-gustavo** [comment]
Build console output for commit: ff3fca3dc345c048009fc8ba03263850734aef9f available at: https://gist.github.com/80b3c838214f61f55a4c830fbc3c981e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7945.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7945.dmg

**Comment by Nex-Pro** [comment]
@Geenz Will do. Will test in localhost due to protocol issue.

**Comment by Nex-Pro** [comment]
#### Fail
**Testing Architecture:** Desktop PC
**CPU:**  4.0 ghz i7
**Ram:** 16gb
**GPU:** Nvidia gtx770 2gb
**Testing OS:** Win 10 64 pro

@Geenz 

Crashes every time I click go home  or type localhost in the address dialog.
Other builds are able to connect to the server of this PR without this issue.
Unable to test at playa due to the protocol change.

**Comment by Geenz** [comment]
Crashes should not be happening as a result of these changes - only a few shaders changed as a result of these changes. Can anyone else reproduce this?

**Comment by Triplelexx** [comment]
Fail

Testing Architecture: Desktop PC
CPU: G328 Pentium
Ram: 8gb
GPU: AMD 270
Testing OS: Win 10 64 home

@Geenz

IT also crashes for me every time I type localhost in the address dialog.

**Comment by Geenz** [comment]
@Nex-Pro I spotted a typo in the shader that may have been causing the crash.  Retest when the build is up.

**Comment by hifi-gustavo** [comment]
Build console output for commit: bd6f6d2eef4916646334b16878bf8cc565c3ce7d available at: https://gist.github.com/80b3c838214f61f55a4c830fbc3c981e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7945.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7945.dmg

**Comment by Geenz** [comment]
I've done some testing with this build - no crashes.  For comparison, I attempted a build prior to this - sure enough it was crashing with the logs complaining about the spot light shader not compiling.  Things *should* be fine now.  Not sure how the typo got in there, likely a result of having one of those gaming mice with the number pad on the side.

**Comment by Triplelexx** [comment]
Pass, the changes have the desired effect with my AMD card. I did notice some banding in the falloff on close inspection, not sure if that's GPU/monitor related. The attenuation is certainly softer and improved with these changes. I'll have to compose a scene later to determine the difference more, but looks good!

**Comment by Triplelexx** [comment]
Hmm, my comment seems to have been lost. Try again.
Pass. 
Crashing resolved, change is visible with my AMD card. I noticed some banding on close inspection but I'm unsure if that's just down to my GPU/monitor. 
I will have to compose a scene later to determine the difference more, certainly noticeable.

**Comment by Nex-Pro** [comment]
### Pass

#7945

**Testing Architecture:** Desktop PC
**CPU:** 4.0 ghz i7
**Ram:** 16gb
**GPU:** Nvidia gtx770 2gb
**Testing OS:** Win 10 64 pro
**Test Location:** localhost server build #7945

Tested only the light entity and settings. Should discuss with @cozza13 if this needs further tests.

@Geenz 
  Works well. I've taken a large number of screen grabs showing settings and effect. Message me if you need them for reference.

zero crashes.

**Comment by Nex-Pro** [comment]
I suspect banding is down to TN screen. Probably far less apparent on an IPS screen.

**Comment by Nex-Pro** [comment]
Triggered a crash on this. Trying to repro.

**Comment by Nex-Pro** [comment]
Crash looks like an object manipulation thing rather than this PR. 

Look at this grab of RGB lights with equal values and same distance above a white primitive.

rgb_falloff

**Comment by Nex-Pro** [comment]
Note how blue looks less intense.

**Comment by Geenz** [comment]
Relative to green, everything we perceive is less intense!

Gotta love how the human eye works, we're most sensitive to green, and least sensitive to blue (due to the fovea not having any blue cones - those are in a ring around the fovea).

**Comment by OmegaHeron** [comment]
I took a look at this just to see how it all looks - not as a QA test.  First, it's a great improvement and - second, I'm happy to see a display where, unless I explicitly make it happen, the primary colors at = intensities do not look equal.  That's > real and part of why imagery can look so un-natural in engines where the R G B values are equalized to appear same intensity across average vision.  Huge :+1: for the way this works.

**Comment by Geenz** [comment]
Not unless there's anything else anyone wants to add, this is pretty much done.

@ZappoMan

**Comment by Nex-Pro** [comment]
@Geenz I am aware of the green thing. I was comparing the blue to both the red and green. :)

**Comment by Nex-Pro** [comment]
@Geenz remember there's no such thing as colour. It's all created in the mind. Which is why I like your orange hair. :P

**Comment by Geenz** [comment]
@Nex-Pro I see infrared people.
(if only)

**Comment by samcake** [comment]
THanks Geenz!

**Comment by Geenz** [comment]
@samcake Not a problem!  Got some time to talk about the RGB10 stuff?  Perhaps via email?

**Comment by samcake** [comment]
@geenz Sorry, ping me directly by email sam@highfidelity.io

**Pull request merged**